### PR TITLE
Add frontend logging functions and log loading time for Dashboard and explore view

### PR DESCRIPTION
**Pull request by graceguo-supercat**
- add frontend logging functions
- add loading log for dash and explore view, also breakdown dash to multiple charts loading time.

This Logger module provides an event-based logging system. Log events are independent of log type (for example page_load_perf). We can define different log type, and attach some new events for other tracking purpose.

Dashboard loading log sample:
Request:
http://localhost:8080/superset/log/?impression_id=rybDJ7xdrf&dashboard_id=35

Post body:
`{`
`source:client`
`type:page_load_perf`
`started_time:1516925660209`
`duration:4539`
`events:{"load_events":[{"label":"slice_44","is_cached":false,"row_count":37,"start_offset":3366,"duration":759},{"label":"slice_240","is_cached":false,"row_count":385,"start_offset":3357,"duration":961},{"label":"slice_53","is_cached":false,"row_count":14,"start_offset":3378,"duration":1143}],"render_events":[{"label":"slice_44","vis_type":"dist_bar","start_offset":4140,"duration":1},{"label":"slice_240","vis_type":"area","start_offset":4327,"duration":0},{"label":"slice_53","vis_type":"big_number","start_offset":4530,"duration":8}]}`
`}`

**source**: for client action logging
**type**: I will use page_load_perf for all dashboard/ explore view page load performance tracking.
**started_time**: timestamp for page load start
**duration**: from page load till all charts are rendered (or error-out)
**events** is serialized json object, which has render_events list and load_events list like this:

```
events: {
    "load_events": [{
        "label": "slice_44",
        "is_cached": false,
        "row_count": 37,
        "start_offset": 3366,
        "duration": 759
    }, {
        "label": "slice_240",
        "is_cached": false,
        "row_count": 385,
        "start_offset": 3357,
        "duration": 961
    }, {
        "label": "slice_53",
        "is_cached": false,
        "row_count": 14,
        "start_offset": 3378,
        "duration": 1143
    }],
    "render_events": [{
        "label": "slice_44",
        "vis_type": "dist_bar",
        "start_offset": 4140,
        "duration": 1
    }, {
        "label": "slice_240",
        "vis_type": "area",
        "start_offset": 4327,
        "duration": 0
    }, {
        "label": "slice_53",
        "vis_type": "big_number",
        "start_offset": 4530,
        "duration": 8
    }]
}
```
@john-bodley @michellethomas @timifasubaa @williaster @fabianmenges

### CODE REVIEW

File: superset/assets/javascripts/chart/chartAction.js [hunk 1]
```diff
@@ -107,15 +108,27 @@ export const RUN_QUERY = 'RUN_QUERY';
...
       .then(() => queryRequest)
-      .then(queryResponse => dispatch(chartUpdateSucceeded(queryResponse, key)))
+      .then((queryResponse) => {
+        Logger.append(LOAD_EVENT, {
+          label: key,
+          isCached: !force,
+          startAt: logStart,
+          duration: Logger.getTimestamp() - logStart,
+        });
+        dispatch(chartUpdateSucceeded(queryResponse, key));
```
Comment (williaster) [inline superset/assets/javascripts/chart/chartAction.js]: I think for redux/promise chaining you need to return the `dispatch`? which this function did before. I think it's not broken now but if you were to chain something else that might not work.

File: superset/assets/javascripts/logger.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    if (handlers[log.eventName].length) {
+      const index = handlers[log.eventName].findIndex(handler => (handler === log.addEvent));
+      handlers[log.eventName].splice(index, 1);
+    }
+  },
+
+  send(log) {
+    const { action, containerId, events, startAt, duration } = log;
+    const url = `/superset/${action}/?${action === LOG_ACTIONS_LOAD_DASHBOARD ? 'dashboard_id' : 'slice_id'}=${containerId}`;
```
Comment (williaster) [inline superset/assets/javascripts/logger.js]: this breaks if there are more than two types of load actions yeah? could we make it check each explicitly?

File: superset/assets/javascripts/explore/components/ExploreViewContainer.jsx [hunk 3]
```diff
@@ -32,18 +33,33 @@ const propTypes = {
...
       height: this.getHeight(),
       width: this.getWidth(),
       showModal: false,
     };
   }
 
+  componentWillMount() {
+    this.actionLog = new ActionLog({
+      action: LOG_ACTIONS_LOAD_EXPLORE,
+      containerId: this.props.slice.slice_id,
```
Comment (williaster) [inline superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: do you think this should just be `id`? it's not exactly a container. cc @john-bodley 

also what if an explore view doesn't have a slice id/isn't saved?

File: superset/assets/javascripts/logger.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    const eventData = {};
+    events.forEach((event) => {
+      const { label, ...eventBody } = event;
```
Comment (williaster) [inline superset/assets/javascripts/logger.js]: it seems like `label` is functioning more like an event name or event key, and all other object keys implicitly get mapped to meta data. could we make the format of the logs more explicit and generic by having only `event` (or `eventName`, to replace `label`) and `meta` keys, where `meta` is an object of anything? 

cc @john-bodley for naming/formatting.

File: superset/assets/javascripts/logger.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import $ from 'jquery';
+
+export const LOG_ACTIONS_LOAD_DASHBOARD = 'load_dashboard';
+export const LOG_ACTIONS_LOAD_EXPLORE = 'load_explore';
+export const LOAD_EVENT = 'load';
```
Comment (williaster) [inline superset/assets/javascripts/logger.js]: it seem strange to add `LOG_ACTIONS_` as a prefix for actions but nothing for event names. what do you think of also adding `LOG_EVENTS_` as a prefix for this?

File: superset/assets/javascripts/chart/chartAction.js (hunk 1 above)
Comment (graceguo-supercat) [reply superset/assets/javascripts/chart/chartAction.js]: Yes. should return dispatch.

File: superset/assets/javascripts/explore/components/ExploreViewContainer.jsx (hunk 3 above)
Comment (graceguo-supercat) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: in existed logging logic, for explore view without existed slice_id, the log entry will have slice_id=0

File: superset/assets/javascripts/logger.js (hunk 5 above)
Comment (graceguo-supercat) [reply superset/assets/javascripts/logger.js]: fixed.

File: superset/assets/javascripts/logger.js (hunk 2 above)
Comment (graceguo-supercat) [reply superset/assets/javascripts/logger.js]: i changed the way tracking type of load actions. i added `source` and `sourceId` attributes in ActionLog instance.

**Comment by timifasubaa** [comment]
PING

**Comment by mistercrunch** [comment]
LGTM

**Comment by graceguo-supercat** [comment]
🎆

**Comment by mistercrunch** [comment]
I think this broke the <CachedLabel/> and  rowcount label. Fix here https://github.com/apache/incubator-superset/pull/4316

**Pull request merged**